### ISSUE TITLE
When running canopy I get TypeError: mkdirp is not a function 

### ISSUE THREAD

**Issue by caw** [jcoglan/canopy#68]
node v20.0.0
npm v9.6.4

MacOS 13.3.1

Canopy installed with 'npm install -g canopy'  - from lib/package.json: ```"version":"0.4.0",```

```
% canopy url.peg --lang python
mkdirp is not a function
TypeError: mkdirp is not a function
```

If I change either:
1:  ``` const mkdirp = require('mkdirp')```

to 

```const {mkdirp} = require('mkdirp')```

**or**  

2: ```await mkdirp(dirname(name))```

to

```await mkdirp.mkdirp(dirname(name))```

all is well. 

I'm sure this is a js question, not really a canopy question, but I'd like to understand what's going on!

Cheers

**Comment by jcoglan** [comment]
This looks like `mkdirp` broke their API again, probably in version 3.0. For now, are you able to install `mkdirp@<3` to work around this?

**Comment by caw** [comment]
For sure - thanks.
Will you update canopy to accommodate that API change? (Asking to learn, I don't have an opinion!)
Cheers

**Comment by caw** [comment]
The python and ruby examples of "implementing actions" have a uniform parameter list
``` self, input, start, end, elements ``` (python) or ```input, start, _end, elements``` (ruby).

But the js actions have different parameters for each action, and make_list seems to receive the first element and then a TreeNode:

```
1 TreeNode {
  text: ',2,3',
  offset: 10,
  elements: [
    TreeNode3 { text: ',2', offset: 10, elements: [Array], value: 2 },
    TreeNode3 { text: ',3', offset: 12, elements: [Array], value: 3 }
  ]
}
```
Why is that? 

And is there a place where the values passed into the parameter list is documented?

thanks for your patience!

Why is JS different?

**Comment by jcoglan** [comment]
I assume you're referring to the examples at https://canopy.jcoglan.com/langs/javascript.html. All the actions `make_map()`, `make_list()`, `make_string()` and `make_number()` take the same arguments -- the input string, the start offset, the end offset, and the matched expression's sub-elements.

In the JS examples the final `elements` parameter is being destructured, i.e. this:

```js
  make_list (input, start, end, [_, first, rest]) {
    rest = [...rest].map((el) => el.value)
    return [first, ...rest]
  }
```

is a shorter way of writing this:

```js
  make_list (input, start, end, elements) {
    let [_, first, rest] = elements
    rest = [...rest].map((el) => el.value)
    return [first, ...rest]
  }
```

`make_list()` is invoked when the expression `"[" value ("," value)* "]"` matched; this expression has four elements:

- `"["` whose result gets bound to `_`
- `value` whose result gets bound to `first`
- `("," value)*` whose result is bound to `rest`
- `"]"` which is ignored by the pattern `[_, first, rest]`; it is fine for a destructuring pattern to have fewer elements than the array it's matching against

**Comment by jcoglan** [comment]
This issue has been fixed by #69.

**Comment by jcoglan** [comment]
This fix is now in release 0.4.1
